src/openforms/contrib/haal_centraal/clients/__init__.py

```diff
@@ -1,25 +1,67 @@
-from zgw_consumers_ext.api_client import ServiceClientFactory
+from typing import Any, cast
 
-from ..models import HaalCentraalConfig
+from openforms.authentication.constants import AuthAttribute
+from openforms.config.models import GlobalConfiguration
+from openforms.submissions.models import Submission
+from zgw_consumers_ext.api_client import build_client
+
+from ..constants import DEFAULT_X_GEBRUIKER_HEADER
+from ..models import HaalCentraalConfig, HaalCentraalRequestOptions
 from .brp import CLIENT_CLS_FOR_VERSION as BRP_CLIENT_CLS_FOR_VERSION, BRPClient
 
 
 class NoServiceConfigured(RuntimeError):
     pass
 
 
-def get_brp_client(**kwargs) -> BRPClient:
+def get_brp_client(submission: Submission, **kwargs: Any) -> BRPClient:
+    """Get an instance of a BRP client.
+
+    :param submission: the current submission in the context of the client usage.
+        This submission will be used to fetch headers overrides.
+    :param kwargs: additional kwargs to be passed to the client class.
+    """
+
     config = HaalCentraalConfig.get_solo()
+    global_config = GlobalConfiguration.get_solo()
     assert isinstance(config, HaalCentraalConfig)
-    if not (service := config.brp_personen_service):
+    assert isinstance(global_config, GlobalConfiguration)
+    if not config.brp_personen_service:
         raise NoServiceConfigured("No BRP service configured!")
 
+    x_headers = {}
+
+    request_options = cast(
+        HaalCentraalRequestOptions | None, submission.form.haal_centraal_request_options
+    )
+    if request_options is not None:
+        x_headers["x-doelbinding"] = request_options.x_doelbinding_header
+        x_headers["x-verwerking"] = request_options.x_verwerking_header
+
+    # Set headers values from the haal centraal config if not set from the form
+    x_headers.setdefault("x-doelbinding", config.default_x_doelbinding_header)
+    x_headers.setdefault("x-verwerking", config.default_x_verwerking_header)
+    if global_config.organization_oin:
+        x_headers["x-origin-oin"] = global_config.organization_oin
+
+    if (
+        submission.registrator
+        and submission.registrator.attribute is AuthAttribute.employee_id
+    ):
+        x_headers["x-gebruiker"] = submission.registrator.value
+    else:
+        x_headers["x-gebruiker"] = DEFAULT_X_GEBRUIKER_HEADER
+
```

Comment:
Not sure where all of this should leave. Either in this helper function or in the factory.

Reply:
I think this is fine, I think you can remove some branches by creating an empty/default instance of the request options if no submission/configuration is present and then simplify lookups like:

```
purpose_limitation = request_options.purpose_limitation or config.default_purpose_limitation
```